### PR TITLE
Major rewrite to simplify new features

### PR DESCRIPTION
**Pull request by erikmol**
I've moved away from the `request.py`/`response.py` to something called `protocol.py` and `protocol.json` to simplify addition of new features. Also moved test cases to a separate folder. 

There are still some things to fix before this PR is complete, but marking it as draft here now for any feedback.

### CODE REVIEW

**Comment by erikmol** [comment]
I've marked this as ready for review now. I've successfully run mower.py with the expected output and run unit tests. I have however not taken any consideration of the code in the HomeAssistant PR but hopefully any changes are straightforward.

**Comment by erikmol** [comment]
Whenever you have time @alistair23 , please review this PR and give some feedback. Long term, it would be great to have some Github Actions running unit tests / linting on PRs as well.

**Comment by alistair23** [comment]
Awesome! This looks great. It's going to take some time for me to get to though. I want to update the current HA pull request first to keep that moving and it looks like this will be a breaking change. Once the HA PR is all fixed up I'll come back to this

File: automower_ble/mower.py [hunk 1]
```diff
@@ -11,375 +11,105 @@
...
-
-        return True
+        This is the same function as get_parameter but with a differerent name to make syntax a bit more clear.
+        It also does not handle any response even though it upstream reads the response."""
+        self.get_parameter(parameter_name, **kwargs)
 
-    def is_connected(self) -> bool:
-        return self.client.is_connected
-
-    async def probe_gatts(self, device):
```
Comment (alistair23) [inline automower_ble/mower.py:217]: We should keep this function. It's used by the HA integration to read information without pairing

File: automower_ble/protocol.py [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+        data[5] = id[1]
+        data[6] = id[2]
+        data[7] = id[3]
+
+        # CRCs and end byte
+        data[9] = crc(data, 1, 8)
+        data.append(crc(data, 1, len(data) - 1))
+        data.append(0x03)
+
+        return data
```
Comment (alistair23) [inline automower_ble/protocol.py:504]: Do you mind running `ruff format` on the PR as well

File: automower_ble/protocol.py [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+
+        if response_data[15] != 0x00:
+            return False
+
+        if response_data[16] != 0x00:
+            return False
+
+        return True
+
+class BLEClient:
```
Comment (alistair23) [inline automower_ble/protocol.py:206]: This feels like it should be in `mower.py` instead of `protocol.py`

**Review by alistair23 — CHANGES_REQUESTED** [review]

File: automower_ble/mower.py (hunk 1 above)
Comment (erikmol) [reply automower_ble/mower.py:217]: It's kept in the parent BLEClient class.

File: automower_ble/protocol.py (hunk 3 above)
Comment (erikmol) [reply automower_ble/protocol.py:225]: I tried to move the code related to protocol and BLE communication to `protocol.py` in order to keep the Mower class smaller just representing and handling the state of the mower. 

Let me know if this makes sense or if you'd like me to move (or rename) BLEClient? Or possibly revert those changes and keep everything in a big Mower class?

**Review by alistair23 — APPROVED** [review]

**Pull request merged**